File: src/api/java/maestro/api/utils/PathCalculationResult.java
package maestro.api.utils;

import java.util.Objects;
import java.util.Optional;
import maestro.api.pathing.calc.IPath;

public class PathCalculationResult {

    private final IPath path;
    private final Type type;

    public PathCalculationResult(Type type) {
        this(type, null);
    }

    public PathCalculationResult(Type type, IPath path) {
        Objects.requireNonNull(type);
        this.path = path;
        this.type = type;
    }

    public final Optional<IPath> getPath() {
        return Optional.ofNullable(this.path);
    }

    public final Type getType() {
        return this.type;
    }

    public enum Type {
        SUCCESS_TO_GOAL,
        SUCCESS_SEGMENT,
        FAILURE,
        CANCELLATION,
        EXCEPTION,
    }
}

File: common/src/main/maestro/utils/PathCalculationResult.kt
package maestro.utils

import maestro.pathing.calc.IPath
import java.util.Optional

/**
 * Represents the result of a pathfinding calculation.
 *
 * Contains an optional path and a result type indicating success, partial success,
 * failure, cancellation, or exception.
 *
 * @property type The result type of the pathfinding calculation
 * @property path The calculated path, if any
 */
data class PathCalculationResult
    @JvmOverloads
    constructor(
        @get:JvmName("getType") val type: Type,
        private val path: IPath? = null,
    ) {
        /**
         * Returns the calculated path wrapped in an Optional.
         *
         * @return Optional containing the path, or empty if no path was calculated
         */
        @JvmName("getPath")
        fun getPath(): Optional<IPath> = Optional.ofNullable(path)

        /**
         * Result types for path calculations.
         */
        enum class Type {
            /** Successfully reached the goal */
            SUCCESS_TO_GOAL,

            /** Successfully calculated a segment towards the goal */
            SUCCESS_SEGMENT,

            /** Failed to find a path */
            FAILURE,

            /** Calculation was cancelled */
            CANCELLATION,

            /** An exception occurred during calculation */
            EXCEPTION,
        }
    }
